packages/chain-keplr-ethereum/src/implementation.ts

```diff
@@ -0,0 +1,172 @@
+import {
+	Action,
+	ActionPayload,
+	Chain,
+	ChainId,
+	ChainImplementation,
+	serializeActionPayload,
+	serializeSessionPayload,
+	Session,
+	SessionPayload,
+} from "@canvas-js/interfaces"
+import { ethers } from "ethers"
+import { configure as configureStableStringify } from "safe-stable-stringify"
+import { decodeSignature, rawSecp256k1PubkeyToRawAddress, Secp256k1Wallet, serializeSignDoc } from "@cosmjs/amino"
+import { Random, Secp256k1, Secp256k1Signature, Sha256 } from "@cosmjs/crypto"
+import { toBech32 } from "@cosmjs/encoding"
+import { KeplrEthereumSigner } from "./signerInterface.js"
+import { getActionSignatureData } from "./signatureData.js"
+
+type Secp256k1WalletPrivateKey = Uint8Array
+
+const sortedStringify = configureStableStringify({
+	bigint: false,
+	circularValue: Error,
+	strict: true,
+	deterministic: true,
+})
+
+const encodeEthAddress = (bech32Prefix: string, address: string) =>
+	toBech32(bech32Prefix, ethers.utils.arrayify(address))
+
+export class KeplrEthereumChainImplementation
+	implements ChainImplementation<KeplrEthereumSigner, Secp256k1WalletPrivateKey>
+{
+	public readonly chain: Chain = "cosmos"
```

Comment:
This introduces some complications because we determine the ChainImplementation to use by matching on `chain`. 

We can either
- Add a different way to determine which ChainImplementation to use by extending `class ChainImplementation`
- Merge the Keplr Ethereum and Terra implementations into the Cosmos chain implementation now, and differentiate between the three inside the implementation methods. This would require us to make the Cosmos chain implementation accept multiple signers as the SessionWallet, i.e. `AminoSigner | KeplrEthereumSigner | EvmMetamaskSigner | TerraSigner`. This seems better to me.
  - On the signature verification side this is easy (see below)
  - On the signing side, we can differentiate by checking the signer the user passes us
  - On delegated signing, the three implementations are mostly the same.

For signature verification, we can use the same code as CW to differentiate between chains:
- (Verifying a session) To determine if a signed session is Terra-based, check the address prefix: `bech32.bech32.decode(sessionPayload.from).prefix === 'terra'`
- (Verifying a session) To determine if a signed session is Ethereum-based: `signature.startsWith('0x')`
- (Verifying an action) Ignore all of the above, actions are signed using a `cosmos` prefix.

For signing, we should mostly rely on the signer the user has passed us, but check that it's not obviously invalid. For example, if we get a Terra signer the chain prefix should always be `terra`.

Reply:
(we discussed this a bit in DMs, repeating it here so that it's on the record)

We have decided to merge the keplr ethereum/terra/evmos implementations into the Cosmos chain implementation. We are using the same logic as CW to differentiate between the signature type when verifying actions/sessions.

Reply:
Looks good!